File: day5/lab/read-mat.f90
PROGRAM matmul

  use mpi

  IMPLICIT NONE

  REAL(kind=8),ALLOCATABLE :: amatrix(:,:), bmatrix(:,:), cmatrix(:,:), buf_mat(:,:)
  REAL(kind=8) :: tmp
  INTEGER :: arows,acols,brows,bcols,crows,ccols
  INTEGER :: i,j,k
  INTEGER :: loc_size_arows, loc_size_brows, loc_size_crows

  INTEGER :: ierror
  INTEGER :: rank, npes, count

  CALL MPI_INIT(ierror)
  CALL MPI_COMM_RANK( MPI_COMM_WORLD, rank, ierror )
  CALL MPI_COMM_SIZE( MPI_COMM_WORLD, npes, ierror )

  ! read dimensions matrix A
  IF( rank == 0 ) THEN 

     open( unit = 20, file = "matrix.dat", status = 'old', form = 'formatted' )  

      READ (20,*) arows,acols

      IF( MOD(arows, npes) /= 0 ) THEN 
         PRINT *, "Dimension no tocmpatible with the number of processes... mod(arows, npes) != 0!" 
         CALL MPI_ABORT( MPI_COMM_WORLD, 1, ierror )
      ENDIF
   ENDIF

   CALL MPI_BCAST( arows, 1, MPI_INTEGER, 0, MPI_COMM_WORLD, ierror )
   CALL MPI_BCAST( acols, 1, MPI_INTEGER, 0, MPI_COMM_WORLD, ierror )

   loc_size_arows = arows / npes
   ALLOCATE( amatrix( loc_size_arows, acols) )

   ! distribute matrix A
   IF ( rank == 0 ) THEN 
      
      ALLOCATE( buf_mat( loc_size_arows, acols) )
      DO i=1,loc_size_arows
         READ (20, *), (amatrix(j,i),j=1,acols)     
      END DO

      do count = 1, npes - 1
         DO i=1,loc_size_arows
            READ (20, *), (buf_mat(j,i),j=1,acols)     
         END DO
         CALL MPI_SEND( buf_mat, loc_size_arows * acols, MPI_DOUBLE_PRECISION, count, 100, MPI_COMM_WORLD, ierror )
      end do
      DEALLOCATE( buf_mat )
      
   ELSE
      CALL MPI_RECV( amatrix, loc_size_arows * acols, MPI_DOUBLE_PRECISION, 0, 100, MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierror )
   ENDIF

  ! read dimensions matrix B
  IF( rank == 0 ) THEN 

      READ (20,*) brows,bcols

      IF( MOD(brows, npes) /= 0 ) THEN 
         PRINT *, "Dimension no tocmpatible with the number of processes... mod(arows, npes) != 0!" 
         CALL MPI_ABORT( MPI_COMM_WORLD, 1, ierror )
      ENDIF
   ENDIF

   CALL MPI_BCAST( brows, 1, MPI_INTEGER, 0, MPI_COMM_WORLD, ierror )
   CALL MPI_BCAST( bcols, 1, MPI_INTEGER, 0, MPI_COMM_WORLD, ierror )

   loc_size_brows = brows / npes
   ALLOCATE( bmatrix( loc_size_brows, bcols) )

   ! distribute matrix B
   IF ( rank == 0 ) THEN 
      
      ALLOCATE( buf_mat( loc_size_brows, bcols) )
      DO i=1,loc_size_brows
         READ (20, *), (bmatrix(j,i),j=1,bcols)
      END DO

      do count = 1, npes - 1
         DO i=1,loc_size_brows
            READ (20, *), (buf_mat(j,i),j=1,bcols)     
         END DO
         CALL MPI_SEND( buf_mat, loc_size_brows * bcols, MPI_DOUBLE_PRECISION, count, 100, MPI_COMM_WORLD, ierror )
      end do
      DEALLOCATE( buf_mat )
      
   ELSE
      CALL MPI_RECV( bmatrix, loc_size_brows * bcols, MPI_DOUBLE_PRECISION, 0, 100, MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierror )
   ENDIF

  crows = loc_size_arows
  ccols = acols
  ALLOCATE(cmatrix(crows,ccols))

  ! IMPLEMENT HERE!!

  CALL MPI_FINALIZE( ierror )

END PROGRAM matmul

File: day5/lab/read-mat.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <mpi.h>

#define TAG 100
#define MPI_PROC_ROOT 0

void read_matrix( double * mat, int n_rows, int n_cols, FILE * fp ){

  int i, j; 

  for( i = 0; i < n_rows; ++i ){
        for( j = 0; j < n_cols; ++j ){
            fscanf( fp, "%lg", &( mat[ i * n_cols + j ] )  );
        }
    }
}

void dist_matrix( double * mat, int n_rows, int n_cols, int rank, int npes, FILE * fp ){

  int count; 
  double * buf_mat = (double *) malloc( n_rows * n_cols * sizeof(double) );  

  if( rank == MPI_PROC_ROOT ){
    read_matrix( mat, n_rows, n_cols, fp );
    for( count = 1; count < npes; count++ ){
      read_matrix( buf_mat, n_rows, n_cols, fp );
      MPI_Send( buf_mat, n_rows * n_cols, MPI_DOUBLE, count, TAG, MPI_COMM_WORLD );
    }
  } 
  else
    MPI_Recv( mat, n_rows * n_cols, MPI_DOUBLE, MPI_PROC_ROOT, TAG, MPI_COMM_WORLD, MPI_STATUS_IGNORE );
  
  free( buf_mat );

}


int main(int argc, char **argv)
{
    double tmp;
    int i,j,k;
    int arows,acols;
    double *amatrix;

    int brows,bcols;
    double *bmatrix;

    int crows,ccols;
    double *cmatrix;

    int rank, npes; 
    int loc_size_arows, loc_size_brows;

    FILE * fp = NULL;

    MPI_Init( &argc, &argv );
    MPI_Comm_rank( MPI_COMM_WORLD, &rank );
    MPI_Comm_size( MPI_COMM_WORLD, &npes) ;

    if( rank == MPI_PROC_ROOT ){ 
      
      fp = fopen( "matrix.dat", "r" );
      
      /* read dimensions matrix A */
      fscanf( fp, "%d%d", &arows, &acols);
      if( arows % npes ){
	fprintf( stderr, "\nDimension no tocmpatible with the number of processes... mod(arows, npes) != 0! \n" );
	MPI_Abort( MPI_COMM_WORLD, 1 );
      }
    }      

    MPI_Bcast( &arows, 1, MPI_INT, MPI_PROC_ROOT, MPI_COMM_WORLD );
    MPI_Bcast( &acols, 1, MPI_INT, MPI_PROC_ROOT, MPI_COMM_WORLD );

    /* read matrix and distribute matrix A */
    loc_size_arows = arows / npes;
    amatrix = (double *) malloc( loc_size_arows * acols * sizeof(double) );
    dist_matrix( amatrix, loc_size_arows, acols, rank, npes, fp );

    if( rank == MPI_PROC_ROOT ){ 

      /* read dimensions matrix B */
      fscanf(fp,"%d%d", &brows, &bcols);
      if( brows % npes ){
	fprintf( stderr, "\nDimension no tocmpatible with the number of processes... mod( brows, npes ) != 0! \n" );
	MPI_Abort( MPI_COMM_WORLD, 1 );
      }

    }      

    MPI_Bcast( &brows, 1, MPI_INT, MPI_PROC_ROOT, MPI_COMM_WORLD );
    MPI_Bcast( &bcols, 1, MPI_INT, MPI_PROC_ROOT, MPI_COMM_WORLD );

    /* read matrix and distribute matrix B */
    loc_size_brows = brows / npes;
    bmatrix = (double *) malloc( loc_size_brows * bcols * sizeof(double) );
    dist_matrix( bmatrix, loc_size_brows, bcols, rank, npes, fp );

    /* allocate matrix C */
    if ((arows != brows) || (acols != bcols)) {
        printf("Matrices not compatible: %d vs. %d, %d vs. %d\n",
               arows, brows, acols, bcols);
        free(amatrix);
        free(bmatrix);
        MPI_Abort( MPI_COMM_WORLD, 1 );
    }

    crows = loc_size_arows;
    ccols = acols;
    cmatrix = (double *) malloc( crows * ccols * sizeof(double) );
    memset( cmatrix, 0, crows * ccols * sizeof(double) );

    /* multiplication */
    /* Implement here the parallel matrix matmul */

    free(amatrix);
    free(bmatrix);
    free(cmatrix);
    if( fp ) fclose( fp );
    
    MPI_Finalize();
    return 0;
}
